### ISSUE TITLE
Unable to start docker service on Win2016 RTM from non-interactive process

### ISSUE THREAD

**Issue by matt-richardson** [docker/docker#27708]
**Description**

I am unable to start the docker service on a Windows Server 2016 RTM instance on AWS (or Azure), from under a non-interactive process.

I am trying to automate some build agents using [packer](packer.io) and as part of that, I'm trying to pre-cache some images. However, the docker service is not started, and I cannot start it.

This also happens when I connect to the instance via a WinRM session, and under a TeamCity session.  

**Steps to reproduce the issue:**
Create a file: `packer.json`, with appropriate values for security group and subnet:

```
{
  "variables": {
    "aws_access_key": "{{env `AWS_ACCESS_KEY`}}",
    "aws_secret_key": "{{env `AWS_SECRET_KEY`}}"
  },
  "builders": [
    {
      "type": "amazon-ebs",
      "access_key": "{{user `aws_access_key`}}",
      "secret_key": "{{user `aws_secret_key`}}",
      "region": "ap-southeast-2",
      "source_ami": "ami-f4774a97",
      "instance_type": "c3.large",
      "ami_name": "2016-with-docker",
      "communicator": "winrm",
      "winrm_username": "Administrator",
      "windows_password_timeout": "20m",
      "subnet_id": "subnet-XXX",
      "associate_public_ip_address": true,
      "security_group_id": "sg-XXX",
      "user_data_file": "setup-winrm.txt"
    }
  ],
  "provisioners": [
    {
      "type":"powershell",
      "inline": [
        "docker version",
        "start-service docker"
      ]
    }
  ]
}
```

create another file: `setup-winrm.txt`:

```
<powershell>
winrm quickconfig -q
winrm set winrm/config/winrs '@{MaxMemoryPerShellMB="2048"}'
winrm set winrm/config '@{MaxTimeoutms="1800000"}'
winrm set winrm/config/service '@{AllowUnencrypted="true"}'
winrm set winrm/config/service/auth '@{Basic="true"}'

netsh advfirewall firewall add rule name="WinRM 5985" protocol=TCP dir=in localport=5985 action=allow
netsh advfirewall firewall add rule name="WinRM 5986" protocol=TCP dir=in localport=5986 action=allow

net stop winrm
& c:\windows\system32\sc.exe config winrm start= auto
net start winrm

Set-ExecutionPolicy -ExecutionPolicy Bypass -Scope LocalMachine
</powershell>
```

execute `packer build --debug packer.json`

**Describe the results you received:**

Service is not started, and does not start:

```
==> amazon-ebs: Prevalidating AMI Name...
==> amazon-ebs: Pausing after run of step 'StepPreValidate'. Press enter to continue. 
==> amazon-ebs: Inspecting the source AMI (ami-f4774a97)...
==> amazon-ebs: Pausing after run of step 'StepSourceAMIInfo'. Press enter to continue. 
==> amazon-ebs: Creating temporary keypair: packer_XXXXX
    amazon-ebs: Saving key for debug purposes: ec2_amazon-ebs.pem
==> amazon-ebs: Pausing after run of step 'StepKeyPair'. Press enter to continue. 
==> amazon-ebs: Pausing after run of step 'StepSecurityGroup'. Press enter to continue. 
==> amazon-ebs: Pausing after run of step 'stepCleanupVolumes'. Press enter to continue. 
==> amazon-ebs: Launching a source AWS instance...
    amazon-ebs: Instance ID: i-XXXXXXXX
==> amazon-ebs: Waiting for instance (i-XXXXXXXX) to become ready...
    amazon-ebs: Public DNS: ec2-XXX-XXX-XXX-XXX.ap-southeast-2.compute.amazonaws.com
    amazon-ebs: Public IP: XXX.XXX.XXX.XXX
    amazon-ebs: Private IP: XXX.XXX.XXX.XXX
==> amazon-ebs: Pausing after run of step 'StepRunSourceInstance'. Press enter to continue. 
==> amazon-ebs: Pausing after run of step 'stepTagEBSVolumes'. Press enter to continue. 
==> amazon-ebs: Waiting for auto-generated password for instance...
    amazon-ebs: It is normal for this process to take up to 15 minutes,
    amazon-ebs: but it usually takes around 5. Please wait.
    amazon-ebs:  
    amazon-ebs: Password retrieved!
    amazon-ebs: Password (since debug is enabled): XXXXXX
==> amazon-ebs: Pausing after run of step 'StepGetPassword'. Press enter to continue. 
==> amazon-ebs: Waiting for WinRM to become available...
==> amazon-ebs: Connected to WinRM!
==> amazon-ebs: Pausing after run of step 'StepConnect'. Press enter to continue. 
==> amazon-ebs: Provisioning with Powershell...
==> amazon-ebs: Provisioning with shell script: /var/folders/3k/dc8q184975n75tv39hb6s0700000gn/T/packer-powershell-provisioner050913598
    amazon-ebs: Client:
    amazon-ebs: Version:      1.12.2-cs2-ws-beta
    amazon-ebs: API version:  1.25
    amazon-ebs: error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.25/version: open //./pipe/docker_engine: The system cannot find the file specified.
    amazon-ebs: Go version:   go1.7.1
    amazon-ebs: Git commit:   050b611
    amazon-ebs: Built:        Tue Oct 11 02:35:40 2016
    amazon-ebs: OS/Arch:      windows/amd64
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: WARNING: Waiting for service 'Docker Engine (docker)' to start...
    amazon-ebs: start-service : Failed to start service 'Docker Engine (docker)'.
    amazon-ebs: At C:\Windows\Temp\script.ps1:2 char:1
    amazon-ebs: + start-service docker
    amazon-ebs: + ~~~~~~~~~~~~~~~~~~~~
    amazon-ebs: + CategoryInfo          : OpenError: (System.ServiceProcess.ServiceController:ServiceController) [Start-Service],
    amazon-ebs: ServiceCommandException
    amazon-ebs: + FullyQualifiedErrorId : StartServiceFailed,Microsoft.PowerShell.Commands.StartServiceCommand
    amazon-ebs:
==> amazon-ebs: Pausing before cleanup of step 'StepConnect'. Press enter to continue. 
```

**Describe the results you expected:**

I expected the service to have already been started / start.

**Additional information you deem important (e.g. issue happens only occasionally):**

I tried this with connecting via WinRM to the launched instance, and I couldn't start the service.
I tried to run `start-service docker` under TeamCity, and it wouldn't start either.
I manually launched an instance, RDP'd to it, disabled the firewall, then WinRM'd to it, and it did work (not sure if the RDP'ing in first did something?). 

**Output of `docker version`:**

```
Client:
Version:      1.12.2-cs2-ws-beta
API version:  1.25
error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.25/version: open //./pipe/docker_engine: The system cannot find the file specified.
Go version:   go1.7.1
Git commit:   050b611
Built:        Tue Oct 11 02:35:40 2016
OS/Arch:      windows/amd64

```

**Output of `docker info`:**

```
PS C:\Users\Administrator\Documents> docker info
docker : Cannot connect to the Docker daemon at npipe:////./pipe/docker_engine. Is the docker daemon running?
    + CategoryInfo          : NotSpecified: (Cannot connect ...daemon running?:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
AWS

**Comment by StefanScherer** [comment]
The service should start from a WinRM session. Check event log or use this script to check typical problems starting Docker service on Win2016/Win10:

```powershell
wget https://raw.githubusercontent.com/Microsoft/Virtualization-Documentation/master/windows-server-container-tools/Debug-ContainerHost/Debug-ContainerHost.ps1 -UseBasicParsing | iex
```

**Comment by matt-richardson** [comment]
Thanks, @StefanScherer. I'd already tried the `Debug-ContainerHost.ps`, but it didn't seem to show much info.

Here's the output:
```
Checking for common problems
start-service : Failed to start service 'Docker Engine (docker)'.
At C:\Windows\Temp\script.ps1:2 char:1
+ start-service docker
+ ~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : OpenError: (System.ServiceProcess.ServiceController:ServiceController) [Start-Service],
ServiceCommandException
+ FullyQualifiedErrorId : StartServiceFailed,Microsoft.PowerShell.Commands.StartServiceCommand

Describing Windows Version and Prerequisites
[+] Is Windows 10 Anniversary Update or Windows Server 2016 652ms
[+] Has KB3192366, KB3194496, or later installed if running Windows build 14393 128ms
[+] Is not a build with blocking issues 22ms
Describing Docker is installed
[+] A Docker service is installed - 'Docker' or 'com.Docker.Service'  736ms
[-] Service is running 73ms
Expected: {Running}
But was:  {Stopped}
30:         $service.Status | Should Be Running
at <ScriptBlock>, <No file>: line 30
[+] Docker.exe is in path 2.46s
error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.25/images/json: open //./pipe/docker_engine: The system cannot find the file specified.
Describing User has permissions to use Docker daemon
[+] docker.exe should not return access denied 33ms
Describing Windows container settings are correct
[+] Do not have DisableVSmbOplock set to 1 31ms
[+] Do not have zz values set 33ms
Describing The right container base images are installed
[-] At least one of 'microsoft/windowsservercore' or 'microsoft/nanoserver' should be installed 82ms
ValidationMetadataException: The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
ParameterBindingValidationException: Cannot validate argument on parameter 'Property'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
at <ScriptBlock>, <No file>: line 87
Showing output from: docker version
Client:
Version:      1.12.2-cs2-ws-beta
API version:  1.25
Go version:   go1.7.1
Git commit:   050b611
Built:        Tue Oct 11 02:35:40 2016
OS/Arch:      windows/amd64

Showing output from: docker info

Warnings & errors from the last 24 hours


ProviderName: docker

TimeCreated                     Id LevelDisplayName Message
-----------                     -- ---------------- -------
10/25/2016 7:52:24 PM            4 Error            fatal: Error starting daemon: Error initializing network control...


Logs saved to C:\Users\Administrator\logs_20161025-195240.csv
```

I'll get more info from that event log message shortly.

**Comment by matt-richardson** [comment]
Event log message is:
```
fatal: Error starting daemon: Error initializing network controller: hnsCall failed in Win32: Server execution failed (0x80080005)
```

**Comment by msabansal** [comment]
Investigating.

**Comment by msabansal** [comment]
We have been able to find the issue. Investigating possible approaches for mitigation. If the user logs on interactively once the problem will not happen again.

**Comment by matt-richardson** [comment]
Hi @msabansal, any progress on this?

**Comment by aidansteele** [comment]
I'm also interested in a solution to this beyond the interactive logon workaround. Our server launches are automated and usually never see a human remote desktop session. We'd be happy in the meantime if there is at least some way to fake an interactive logon so our work can be unblocked by this.

**Comment by msabansal** [comment]
Appologies for getting back so late on this. For now as a workaround you can delete the Registry key in HKLM:\SYSTEM\CurrentControlSet\Control\Wininit\Headless, 

We are releasing a fix for this as an update this week.

**Comment by aidansteele** [comment]
Thanks @msabansal - hugely appreciated! Just as a clarification, is the fix being released for Docker or for Windows?

**Comment by friism** [comment]
@matt-richardson unrelated, but you might be interested in [this experiment](https://github.com/friism/ec2-windows-packer-ssh) using SSH instead of WinRM to configure and access EC2 Windows instances. There's actually a similar problem because a password login is required to load the user profile

**Comment by matt-richardson** [comment]
I can confirm that
```PowerShell
Remove-ItemProperty -Path "HKLM:\SYSTEM\CurrentControlSet\Control\Wininit" -Name "Headless"
```

Works successfully.

Looking forward to the proper fix though.

**Comment by matt-richardson** [comment]
Hi @msabansal - did that fix get released? Also,  as @aidansteele asked, is this a fix in Windows or in Docker?

**Comment by heynickc** [comment]
Hi all, I'm having the same symptoms as @matt-richardson, (`fatal: Error starting daemon: Error initializing network controller: hnsCall failed in Win32: Server execution failed (0x80080005)"`) except I'm using TeamCity on Azure and trying to start the docker service using a FAKE build script.  I tried adding the registry key fix to the build script, preceding the attempt to re-start the service.  Still the same symptom.  Here's the output:

```
[11:42:34][Step 1/1] Target: RestartDocker
[11:42:34][Step 1/1] powershell.exe Remove-ItemProperty -Path 'HKLM:\SYSTEM\CurrentControlSet\Control\Wininit' -Name 'Headless'
[11:42:37][Step 1/1] Starting Service docker on PBW1537
[11:42:37][Step 1/1] Waiting for docker to start (Timeout: 00:02:00)
[11:42:38][Step 1/1] Waiting for docker to start (Timeout: 00:02:00)
[11:42:39][Step 1/1] Waiting for docker to start (Timeout: 00:02:00)
[11:42:40][Step 1/1] Waiting for docker to start (Timeout: 00:02:00)
[11:42:41][Step 1/1] Waiting for docker to start (Timeout: 00:02:00)
[11:42:42][Step 1/1] Waiting for docker to start (Timeout: 00:02:00)
[11:42:43][Step 1/1] Waiting for docker to start (Timeout: 00:02:00)
[11:42:44][Step 1/1] Waiting for docker to start (Timeout: 00:02:00)
[11:42:45][Step 1/1] Waiting for docker to start (Timeout: 00:02:00)
[11:42:46][Step 1/1] Waiting for docker to start (Timeout: 00:02:00)

... continue until timeout
```

I'm willing to automate some sort of login if I have to, so I guess my question is: what would be an alternative if that registry key fix doesn't work?  Since the Azure VMs are built on the fly, I need to do my best to try to fix the existing VM.  Versus deploy a new VM, RDP into it, fix it, then re-image it, then deploy that to TeamCity, etc...

Here's my `docker version`.  On the TeamCity-initialized VM, can't get the `Server:...` version since it's not started, however in an RDP session it looks like this:

```
Client:
 Version:      1.12.2-cs2-ws-beta
 API version:  1.25
 Go version:   go1.7.1
 Git commit:   050b611
 Built:        Tue Oct 11 02:35:40 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.12.2-cs2-ws-beta
 API version:  1.25
 Go version:   go1.7.1
 Git commit:   050b611
 Built:        Tue Oct 11 02:35:40 2016
 OS/Arch:      windows/amd64
```

If I'm behind on Docker version, can I update it in a "headless" fashion?  Thanks in advance for your help.

**Comment by msabansal** [comment]
The 12B package for windows has the fix for this issue. 

@heynickc You will need to reboot the system once for the fix to work. I think Azure container images will be updated with the fix.

**Comment by matt-richardson** [comment]
@msabansal Can you be more specific? The 12B package of what?

**Comment by msabansal** [comment]
@matt-richardson 12B is the next windows update. It should be already available if you update your windows build

**Comment by heynickc** [comment]
Thank you @msabansal.  Just so I'm clear, the update should be included in one of the following?

```
sconfig "6" then "A"...

Microsoft (R) Windows Script Host Version 5.812
Copyright (C) Microsoft Corporation. All rights reserved.

Search for for (A)ll updates or (R)ecommended updates only? A

Searching for all applicable updates...

List of applicable items on the machine:

1> Update for Windows Server 2016 for x64-based Systems (KB3199986)
2> Windows Malicious Software Removal Tool for Windows Insider Preview and Server Technical Preview x64 - December 2016 (KB890830)
3> Windows Malicious Software Removal Tool for Windows 8, 8.1, 10 and Windows Server 2012, 2012 R2, 2016 x64 Edition - December 2016 (KB890830)
4> Cumulative Update for Windows Server 2016 for x64-based Systems (KB3206632)
5> Definition Update for Windows Defender - KB2267602 (Definition 1.233.2789.0)

Select an option:
(A)ll updates, (N)o updates or (S)elect a single update?
```

**Comment by JMesser81** [comment]
@heynickc - "12B" simply means the December (12) second week (B) update. This corresponds to KB3206632.
https://support.microsoft.com/en-us/help/4004227

**Comment by heynickc** [comment]
Thanks @JMesser81.  Installed update and confirmed issue is resolved

**Comment by thaJeztah** [comment]
@matt-richardson can you confirm if the latest Windows update resolves this issue for you as well?

**Comment by thaJeztah** [comment]
I'll go ahead and close

**Comment by matt-richardson** [comment]
Apologies for the delay in responding, @thaJeztah.

I've just removed the fix (mentioned [above](https://github.com/docker/docker/issues/27708#issuecomment-262113883) and tested against a recent AWS base image (ami-7e70741d), and it succeeds as expected.

It appears that the windows update fixes are cumulative, and are not listed individually anyway, so its very difficult to figure out if you've actually got KB3206632 installed or not. Going by [this page](https://support.microsoft.com/en-au/help/4004227/windows-10-update-kb3206632), KB3206632 sets the OS version to `14393.576`, and my testing was done against OS version `14393.693`, so I presume I've got it installed.

Thanks all for investigating and fixing.
